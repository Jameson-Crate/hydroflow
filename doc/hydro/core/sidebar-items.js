window.SIDEBAR_ITEMS = {"enum":["ClientStrategy","HostTargetType","ServerStrategy"],"mod":["custom_service","deployment","gcp","hydroflow_crate","localhost","terraform","util"],"struct":["ResourceBatch","ResourcePool","ResourceResult"],"trait":["Host","LaunchedBinary","LaunchedHost","Service"]};