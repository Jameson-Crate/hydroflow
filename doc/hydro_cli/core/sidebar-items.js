window.SIDEBAR_ITEMS = {"enum":[["ConnectionType",""]],"mod":[["deployment",""],["hydroflow_crate",""],["localhost",""]],"struct":[["ResourceBatch",""],["ResourceResult",""]],"trait":[["Host",""],["LaunchedBinary",""],["LaunchedHost",""],["Service",""]]};