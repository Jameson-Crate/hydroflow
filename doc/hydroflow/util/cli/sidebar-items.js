window.SIDEBAR_ITEMS = {"enum":["BindConfig","BoundConnection","ConnectionPipe"],"fn":["init"],"struct":["ConnectedBidi","ConnectedDemux"],"trait":["Connected"]};