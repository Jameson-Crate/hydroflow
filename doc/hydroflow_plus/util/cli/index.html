<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `cli` mod in crate `hydroflow_plus`."><title>hydroflow_plus::util::cli - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../static.files/rustdoc-e935ef01ae1c1829.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="hydroflow_plus" data-themes="" data-resource-suffix="" data-rustdoc-version="1.78.0-nightly (8f359beca 2024-02-23)" data-channel="nightly" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../../static.files/storage-4c98445ec4002617.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-c37d3936c59ababd.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../hydroflow_plus/index.html">hydroflow_plus</a><span class="version">0.6.0</span></h2></div><h2 class="location"><a href="#">Module cli</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Aliases</a></li></ul></section><h2><a href="../index.html">In hydroflow_plus::util</a></h2></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../../hydroflow_plus/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../index.html">hydroflow_plus</a>::<wbr><a href="../index.html">util</a>::<wbr><a class="mod" href="#">cli</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../../../src/hydroflow/util/mod.rs.html#29">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ConnectedDemux.html" title="struct hydroflow_plus::util::cli::ConnectedDemux">ConnectedDemux</a></div></li><li><div class="item-name"><a class="struct" href="struct.ConnectedDirect.html" title="struct hydroflow_plus::util::cli::ConnectedDirect">ConnectedDirect</a></div></li><li><div class="item-name"><a class="struct" href="struct.ConnectedTagged.html" title="struct hydroflow_plus::util::cli::ConnectedTagged">ConnectedTagged</a></div></li><li><div class="item-name"><a class="struct" href="struct.DemuxDrain.html" title="struct hydroflow_plus::util::cli::DemuxDrain">DemuxDrain</a></div></li><li><div class="item-name"><a class="struct" href="struct.HydroCLI.html" title="struct hydroflow_plus::util::cli::HydroCLI">HydroCLI</a></div></li><li><div class="item-name"><a class="struct" href="struct.MergeSource.html" title="struct hydroflow_plus::util::cli::MergeSource">MergeSource</a></div></li><li><div class="item-name"><a class="struct" href="struct.TaggedSource.html" title="struct hydroflow_plus::util::cli::TaggedSource">TaggedSource</a></div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.BoundConnection.html" title="enum hydroflow_plus::util::cli::BoundConnection">BoundConnection</a></div></li><li><div class="item-name"><a class="enum" href="enum.RealizedServerPort.html" title="enum hydroflow_plus::util::cli::RealizedServerPort">RealizedServerPort</a></div></li><li><div class="item-name"><a class="enum" href="enum.ServerBindConfig.html" title="enum hydroflow_plus::util::cli::ServerBindConfig">ServerBindConfig</a></div></li><li><div class="item-name"><a class="enum" href="enum.ServerOrBound.html" title="enum hydroflow_plus::util::cli::ServerOrBound">ServerOrBound</a></div></li><li><div class="item-name"><a class="enum" href="enum.ServerPort.html" title="enum hydroflow_plus::util::cli::ServerPort">ServerPort</a></div><div class="desc docblock-short">Describes how to connect to a service which is listening on some port.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.Connected.html" title="trait hydroflow_plus::util::cli::Connected">Connected</a></div></li><li><div class="item-name"><a class="trait" href="trait.ConnectedSink.html" title="trait hydroflow_plus::util::cli::ConnectedSink">ConnectedSink</a></div></li><li><div class="item-name"><a class="trait" href="trait.ConnectedSource.html" title="trait hydroflow_plus::util::cli::ConnectedSource">ConnectedSource</a></div></li><li><div class="item-name"><a class="trait" href="trait.StreamSink.html" title="trait hydroflow_plus::util::cli::StreamSink">StreamSink</a></div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.init.html" title="fn hydroflow_plus::util::cli::init">init</a></div></li><li><div class="item-name"><a class="fn" href="fn.launch.html" title="fn hydroflow_plus::util::cli::launch">launch</a></div></li><li><div class="item-name"><a class="fn" href="fn.launch_flow.html" title="fn hydroflow_plus::util::cli::launch_flow">launch_flow</a></div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.BufferedDrain.html" title="type hydroflow_plus::util::cli::BufferedDrain">BufferedDrain</a></div></li><li><div class="item-name"><a class="type" href="type.DynSink.html" title="type hydroflow_plus::util::cli::DynSink">DynSink</a></div></li><li><div class="item-name"><a class="type" href="type.DynStream.html" title="type hydroflow_plus::util::cli::DynStream">DynStream</a></div></li><li><div class="item-name"><a class="type" href="type.DynStreamSink.html" title="type hydroflow_plus::util::cli::DynStreamSink">DynStreamSink</a></div></li><li><div class="item-name"><a class="type" href="type.InitConfig.html" title="type hydroflow_plus::util::cli::InitConfig">InitConfig</a></div></li></ul></section></div></main></body></html>