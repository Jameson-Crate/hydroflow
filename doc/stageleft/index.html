<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `stageleft` crate."><title>stageleft - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="stageleft" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.0-nightly (bf3c6c5be 2024-02-01)" data-channel="nightly" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-4c98445ec4002617.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../stageleft/index.html">stageleft</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><ul class="block">
            <li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../stageleft/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">stageleft</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/stageleft/lib.rs.html#1-248">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="internal/index.html" title="mod stageleft::internal">internal</a></div></li><li><div class="item-name"><a class="mod" href="runtime_support/index.html" title="mod stageleft::runtime_support">runtime_support</a></div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.PATH_SEPARATOR.html" title="macro stageleft::PATH_SEPARATOR">PATH_SEPARATOR</a></div></li><li><div class="item-name"><a class="macro" href="macro.q.html" title="macro stageleft::q">q</a></div><div class="desc docblock-short">Creates a quoted expression, which can be typechecked but has its AST serialized
until it is spliced into a staged entrypoint. Returns a value which implements
<code>Quoted&lt;T&gt;</code>, where <code>T</code> is the type of the expression, and also may implement
<code>IntoQuotedMut&lt;T&gt;</code> is the values it captures would be safe to capture inside an
<code>FnMut</code> context.</div></li><li><div class="item-name"><a class="macro" href="macro.quse_fn.html" title="macro stageleft::quse_fn">quse_fn</a></div></li><li><div class="item-name"><a class="macro" href="macro.stageleft_crate.html" title="macro stageleft::stageleft_crate">stageleft_crate</a></div></li><li><div class="item-name"><a class="macro" href="macro.stageleft_macro_crate.html" title="macro stageleft::stageleft_macro_crate">stageleft_macro_crate</a></div></li><li><div class="item-name"><a class="macro" href="macro.stageleft_no_entry_crate.html" title="macro stageleft::stageleft_no_entry_crate">stageleft_no_entry_crate</a></div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.RuntimeData.html" title="struct stageleft::RuntimeData">RuntimeData</a></div><div class="desc docblock-short">Represents a piece of data that will be passed into the generated code</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.IntoQuotedMut.html" title="trait stageleft::IntoQuotedMut">IntoQuotedMut</a></div></li><li><div class="item-name"><a class="trait" href="trait.IntoQuotedOnce.html" title="trait stageleft::IntoQuotedOnce">IntoQuotedOnce</a></div></li><li><div class="item-name"><a class="trait" href="trait.Quoted.html" title="trait stageleft::Quoted">Quoted</a></div></li><li><div class="item-name"><a class="trait" href="trait.QuotedContext.html" title="trait stageleft::QuotedContext">QuotedContext</a></div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.entry.html" title="attr stageleft::entry">entry</a></div><div class="desc docblock-short">Defines an entrypoint for staged code, which will be available as a proc macro.
The entrypoint must be a function that returns <code>impl Quoted&lt;T&gt;</code> for some type <code>T</code>.</div></li><li><div class="item-name"><a class="attr" href="attr.runtime.html" title="attr stageleft::runtime">runtime</a></div><div class="desc docblock-short">Marks a chunk of code as being runtime-only, which means that no staged code in its crate can depend on it.
Code behind this attribute is allowed to use staged entrypoints defined in the same crate.</div></li></ul></section></div></main></body></html>