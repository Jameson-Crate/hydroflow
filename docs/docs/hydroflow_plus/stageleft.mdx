---
title: Stageleft
sidebar_position: 4
---

import StageleftDocs from '../../../stageleft/README.md'

<StageleftDocs/>
