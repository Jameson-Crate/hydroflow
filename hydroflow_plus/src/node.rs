use std::cell::RefCell;
use std::rc::Rc;

use hydroflow::util::cli::HydroCLI;
use stageleft::RuntimeData;

pub trait HfNode<'a>: Clone {
    fn id(&self) -> usize;
    fn next_port(&self) -> String;
    fn get_cli(&self) -> RuntimeData<&'a HydroCLI>;
}

impl<'a> HfNode<'a> for () {
    fn id(&self) -> usize {
        0
    }

    fn next_port(&self) -> String {
        panic!();
    }

    fn get_cli(&self) -> RuntimeData<&'a HydroCLI> {
        panic!();
    }
}

impl<'a> HfNode<'a> for usize {
    fn id(&self) -> usize {
        *self
    }

    fn next_port(&self) -> String {
        panic!();
    }

    fn get_cli(&self) -> RuntimeData<&'a HydroCLI> {
        panic!();
    }
}

#[derive(Clone)]
pub struct CLIRuntimeNode<'a> {
    id: usize,
    next_port: Rc<RefCell<usize>>,
    cli: RuntimeData<&'a HydroCLI>,
}

impl<'a> CLIRuntimeNode<'a> {
    pub fn new(id: usize, cli: RuntimeData<&'a HydroCLI>) -> CLIRuntimeNode {
        CLIRuntimeNode {
            id,
            next_port: Rc::new(RefCell::new(0)),
            cli,
        }
    }
}

impl<'a> HfNode<'a> for CLIRuntimeNode<'a> {
    fn id(&self) -> usize {
        self.id
    }

    fn next_port(&self) -> String {
        let next_send_port = *self.next_port.borrow();
        *self.next_port.borrow_mut() += 1;
        format!("port_{}", next_send_port)
    }

    fn get_cli(&self) -> RuntimeData<&'a HydroCLI> {
        self.cli
    }
}

pub trait HfConnectable<'a, O: HfNode<'a>> {
    fn connect(&self, other: &O, source_port: &str, recipient_port: &str);
}

impl<'a> HfConnectable<'a, CLIRuntimeNode<'a>> for CLIRuntimeNode<'a> {
    fn connect(&self, _other: &CLIRuntimeNode, _source_port: &str, _recipient_port: &str) {}
}

pub trait HfNodeBuilder<N> {
    fn build(&mut self, id: usize) -> N;
}

pub struct CLIRuntimeNodeBuilder<'a> {
    cli: RuntimeData<&'a HydroCLI>,
}

impl CLIRuntimeNodeBuilder<'_> {
    pub fn new(cli: RuntimeData<&HydroCLI>) -> CLIRuntimeNodeBuilder {
        CLIRuntimeNodeBuilder { cli }
    }
}

impl<'a> HfNodeBuilder<CLIRuntimeNode<'a>> for CLIRuntimeNodeBuilder<'a> {
    fn build(&mut self, id: usize) -> CLIRuntimeNode<'a> {
        CLIRuntimeNode::new(id, self.cli)
    }
}

pub trait HFDeploy<'a> {
    type Node: HfNode<'a> + HfConnectable<'a, Self::Node>;
    type NodeBuilder: HfNodeBuilder<Self::Node>;
}

pub struct CLIRuntime<'b> {
    _marker: std::marker::PhantomData<&'b ()>,
}

impl<'a: 'b, 'b> HFDeploy<'a> for CLIRuntime<'b> {
    type Node = CLIRuntimeNode<'a>;
    type NodeBuilder = CLIRuntimeNodeBuilder<'a>;
}
