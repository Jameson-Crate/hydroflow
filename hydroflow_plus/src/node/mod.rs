use std::cell::RefCell;
use std::io;
use std::marker::PhantomData;

use hydroflow::bytes::BytesMut;
use hydroflow::futures::stream::Stream as FuturesStream;
use proc_macro2::Span;
use stageleft::Quoted;
use syn::parse_quote;

use crate::builder::Builders;
use crate::stream::{Async, Windowed};
use crate::{GraphBuilder, HfCycle, Stream};

mod graphs;
pub use graphs::*;

pub mod network;
pub use network::*;

pub trait LocalDeploy<'a> {
    type Node: HfNode<'a, Meta = Self::Meta>;
    type Cluster: HfNode<'a, Meta = Self::Meta> + HfCluster<'a>;
    type Meta: Default;
    type RuntimeID;
}

pub trait Deploy<'a> {
    type Node: HfNode<'a, Meta = Self::Meta, Port = Self::NodePort>
        + HfSendOneToOne<'a, Self::Node>
        + HfSendOneToMany<'a, Self::Cluster>;
    type Cluster: HfNode<'a, Meta = Self::Meta, Port = Self::ClusterPort>
        + HfSendManyToOne<'a, Self::Node>
        + HfSendManyToMany<'a, Self::Cluster>
        + HfCluster<'a>;
    type NodePort;
    type ClusterPort;
    type Meta: Default;
    type RuntimeID;
}

impl<
        'a,
        T: Deploy<'a, Node = N, Cluster = C, Meta = M, RuntimeID = R>,
        N: HfNode<'a, Meta = M> + HfSendOneToOne<'a, N> + HfSendOneToMany<'a, C>,
        C: HfNode<'a, Meta = M> + HfSendManyToOne<'a, N> + HfSendManyToMany<'a, C> + HfCluster<'a>,
        M: Default,
        R,
    > LocalDeploy<'a> for T
{
    type Node = N;
    type Cluster = C;
    type Meta = M;
    type RuntimeID = R;
}

pub trait NodeBuilder<'a, D: LocalDeploy<'a> + ?Sized> {
    fn build(&self, id: usize, builder: &'a GraphBuilder<'a, D>, meta: &mut D::Meta) -> D::Node;
}

pub trait ClusterBuilder<'a, D: LocalDeploy<'a> + ?Sized> {
    fn build(&self, id: usize, builder: &'a GraphBuilder<'a, D>, meta: &mut D::Meta) -> D::Cluster;
}

pub trait HfNode<'a>: Clone {
    type Port;
    type Meta;

    fn id(&self) -> usize;
    fn graph_builder(&self) -> (&'a RefCell<usize>, &'a Builders);
    fn next_port(&self) -> Self::Port;

    fn update_meta(&mut self, meta: &Self::Meta);

    fn source_stream<T, E: FuturesStream<Item = T> + Unpin>(
        &self,
        e: impl Quoted<'a, E>,
    ) -> Stream<'a, T, Async, Self> {
        let (next_id_cell, builders) = self.graph_builder();

        let next_id = {
            let mut next_id = next_id_cell.borrow_mut();
            let id = *next_id;
            *next_id += 1;
            id
        };

        let ident = syn::Ident::new(&format!("stream_{}", next_id), Span::call_site());
        let e = e.splice();

        builders
            .borrow_mut()
            .as_mut()
            .unwrap()
            .entry(self.id())
            .or_default()
            .add_statement(parse_quote! {
                #ident = source_stream(#e) -> tee();
            });

        Stream {
            ident,
            node: self.clone(),
            next_id: next_id_cell,
            builders,
            _phantom: PhantomData,
        }
    }

    fn source_external(
        &self,
    ) -> (
        Self::Port,
        Stream<'a, Result<BytesMut, io::Error>, Async, Self>,
    )
    where
        Self: HfSendOneToOne<'a, Self>,
    {
        let (next_id_cell, builders) = self.graph_builder();

        let next_id = {
            let mut next_id = next_id_cell.borrow_mut();
            let id = *next_id;
            *next_id += 1;
            id
        };

        let ident = syn::Ident::new(&format!("stream_{}", next_id), Span::call_site());
        let port = self.next_port();
        let source_pipeline = Self::gen_source_statement(self, &port);

        builders
            .borrow_mut()
            .as_mut()
            .unwrap()
            .entry(self.id())
            .or_default()
            .add_statement(parse_quote! {
                #ident = #source_pipeline -> tee();
            });

        (
            port,
            Stream {
                ident,
                node: self.clone(),
                next_id: next_id_cell,
                builders,
                _phantom: PhantomData,
            },
        )
    }

    fn source_iter<T, E: IntoIterator<Item = T>>(
        &self,
        e: impl Quoted<'a, E>,
    ) -> Stream<'a, T, Windowed, Self> {
        let (next_id_cell, builders) = self.graph_builder();

        let next_id = {
            let mut next_id = next_id_cell.borrow_mut();
            let id = *next_id;
            *next_id += 1;
            id
        };

        let ident = syn::Ident::new(&format!("stream_{}", next_id), Span::call_site());
        let e = e.splice();

        builders
            .borrow_mut()
            .as_mut()
            .unwrap()
            .entry(self.id())
            .or_default()
            .add_statement(parse_quote! {
                #ident = source_iter(#e) -> tee();
            });

        Stream {
            ident,
            node: self.clone(),
            next_id: next_id_cell,
            builders,
            _phantom: PhantomData,
        }
    }

    fn cycle<T, W>(&self) -> (HfCycle<'a, T, W, Self>, Stream<'a, T, W, Self>) {
        let (next_id_cell, builders) = self.graph_builder();

        let next_id = {
            let mut next_id = next_id_cell.borrow_mut();
            let id = *next_id;
            *next_id += 1;
            id
        };

        let ident = syn::Ident::new(&format!("stream_{}", next_id), Span::call_site());

        builders
            .borrow_mut()
            .as_mut()
            .unwrap()
            .entry(self.id())
            .or_default()
            .add_statement(parse_quote! {
                #ident = tee();
            });

        (
            HfCycle {
                ident: ident.clone(),
                node: self.clone(),
                builders,
                _phantom: PhantomData,
            },
            Stream {
                ident,
                node: self.clone(),
                next_id: next_id_cell,
                builders,
                _phantom: PhantomData,
            },
        )
    }
}

pub trait HfCluster<'a> {
    fn ids<'b>(&'b self) -> impl Quoted<'a, &'a Vec<u32>> + Copy + 'a;
}
