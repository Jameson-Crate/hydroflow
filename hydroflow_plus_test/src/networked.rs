use hydroflow::bytes::BytesMut;
use hydroflow::util::cli::HydroCLI;
use hydroflow_plus::node::{HfNetworkedDeploy, HfNode, HfNodeBuilder};
use hydroflow_plus::scheduled::graph::Hydroflow;
use hydroflow_plus::HfBuilder;
use hydroflow_plus_cli_integration::{CLIRuntime, HydroflowPlusMeta};
use stageleft::{q, Quoted, RuntimeData};

pub fn networked_basic<'a, D: HfNetworkedDeploy<'a>>(
    graph: &'a HfBuilder<'a, D>,
    node_builder: &impl HfNodeBuilder<'a, D>,
) -> (D::NodePort, D::Node, D::Node) {
    let node_zero = graph.node(node_builder);
    let node_one = graph.node(node_builder);

    let (source_zero_port, source_zero) = node_zero.source_external();

    source_zero
        .map(q!(|v| v.unwrap().freeze()))
        .send_bytes(&node_one)
        .for_each(q!(|v: Result<BytesMut, _>| {
            println!(
                "node one received: {:?}",
                std::str::from_utf8(&v.unwrap()).unwrap()
            );
        }));

    (source_zero_port, node_zero, node_one)
}

#[stageleft::entry]
pub fn networked_basic_runtime<'a>(
    graph: &'a HfBuilder<'a, CLIRuntime>,
    cli: RuntimeData<&'a HydroCLI<HydroflowPlusMeta>>,
    node_id: RuntimeData<usize>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let _ = networked_basic(graph, &cli);
    graph.build(node_id)
}

#[stageleft::runtime]
#[cfg(test)]
mod tests {
    use std::time::Duration;

    use hydro_cli::core::Deployment;
    use hydroflow::futures::SinkExt;
    use hydroflow::util::cli::ConnectedSink;
    use hydroflow_plus_cli_integration::{CLIDeployNodeBuilder, DeployCrateWrapper};

    #[tokio::test]
    async fn networked_basic() {
        let mut deployment = Deployment::new();
        let localhost = deployment.Localhost();

        let builder = hydroflow_plus::HfBuilder::new();
        let (source_zero_port, _, node_one) = super::networked_basic(
            &builder,
            &CLIDeployNodeBuilder::new(|id| {
                deployment.HydroflowCrate(
                    ".",
                    localhost.clone(),
                    Some("networked_basic".into()),
                    None,
                    Some("dev".into()),
                    None,
                    Some(vec![id.to_string()]),
                    None,
                    vec![],
                )
            }),
        );
        builder.wire();

        let port_to_zero = source_zero_port
            .create_sender(&mut deployment, &localhost)
            .await;

        deployment.deploy().await.unwrap();

        let mut conn_to_zero = port_to_zero.connect().await.into_sink();
        let node_one_stdout = node_one.stdout().await;

        deployment.start().await.unwrap();

        conn_to_zero.send("hello world!".into()).await.unwrap();

        assert_eq!(
            tokio::time::timeout(Duration::from_secs(1), node_one_stdout.recv())
                .await
                .unwrap()
                .unwrap(),
            "node one received: \"hello world!\""
        );
    }
}
