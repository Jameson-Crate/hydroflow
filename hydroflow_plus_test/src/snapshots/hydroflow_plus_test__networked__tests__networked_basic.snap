---
source: hydroflow_plus_test/src/networked.rs
expression: builder.ir()
---
[
    ForEach {
        f: { use crate :: __staged :: networked :: * ; | v : Result < BytesMut , _ > | { println ! ("node one received: {:?}" , std :: str :: from_utf8 (& v . unwrap ()) . unwrap ()) ; } },
        input: Network {
            to_location: 1,
            serialize_pipeline: None,
            sink_expr: null,
            source_expr: null,
            deserialize_pipeline: None,
            input: Map {
                f: { use crate :: __staged :: networked :: * ; | v | v . unwrap () . freeze () },
                input: Source {
                    source: Stream(
                        null,
                    ),
                    produces_delta: false,
                    location_id: 0,
                },
            },
        },
    },
    ForEach {
        f: { use crate :: __staged :: networked :: * ; | v : Result < BytesMut , _ > | { println ! ("cluster received: {:?}" , std :: str :: from_utf8 (& v . unwrap ()) . unwrap ()) ; } },
        input: Source {
            source: Stream(
                null,
            ),
            produces_delta: false,
            location_id: 2,
        },
    },
]
